[ONNX] Export module as function

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105618

Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

BowenBao committed:
[ONNX] Export module as function

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -634,9 +675,82 @@ def call_method(self, node: torch.fx.Node):
+        # TODO: We may want to consider other naming styles. The goal is to be stable and
+        # unique such that it can be easily identified in case of kernel substitution.
+        # Example for current style is combination of qualified module class name and
+        # module attribute name: `torch_nn_modules_conv_Conv2d_conv1)`.
```
nit )

BowenBao committed:
forgot to push the main thing on "[ONNX] Export module as function"




[ghstack-poisoned]

justinchuby commented:
Is the pass always turned on?

justinchuby commented on .ci/docker/common/install_onnx.sh:
```diff
@@ -12,8 +12,18 @@ pip_install \
+# # - Segfault fix for functions.
+# # - Duplicated node name fix for function inlining.
+# pip install coloredlogs  # dependency of ort-nightly
+# pip_install --index-url=https://aiinfra.pkgs.visualstudio.com/PublicPackages/_packaging/ORT-Nightly/pypi/simple/ ort-nightly==1.16.0.dev20230528001
```
If we used this do we still need to inline? Good to not install from ado though I remember it can be unstable

BowenBao replied:
Still need inlining, unless onnx submodule is updated in ORT.

I don't remember if we need this even with inline though, so leaving it here for CI to do a quick check.

justinchuby commented on .ci/docker/common/install_onnx.sh:
```diff
@@ -12,8 +12,18 @@ pip_install \
+# Using 1.15dev branch for the following not yet released features and fixes.
+# - Segfault fix for shape inference.
+# - Inliner to workaround ORT segfault.
+pip_install "onnx@git+https://github.com/onnx/onnx@213b525a51ead28961d9b4f2764b08c7e336bf2c"
```
Install from onnx-weekly?

BowenBao replied:
good idea

BowenBao replied:
```
2023-07-20T01:48:31.9917154Z Installing collected packages: onnx-weekly
2023-07-20T01:48:31.9917755Z Successfully installed onnx-weekly-1.15.0.dev20230717
...
2023-07-20T01:48:50.4654837Z Collecting onnx>=1.14 (from onnxscript@ git+https://github.com/microsoft/onnxscript@2bb3e9f2d094912f81cb63cecb412efb14c65738)
2023-07-20T01:48:50.4655481Z   Downloading onnx-1.14.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (14.6 MB)
...
2023-07-20T01:48:50.4659645Z Successfully built onnxscript
2023-07-20T01:48:50.4659894Z Installing collected packages: onnx, onnxscript
2023-07-20T01:48:50.4660249Z Successfully installed onnx-1.14.0 onnxscript-0.1.0.dev20230720
```
@justinchuby onnx-weekly not recognized by onnxscript, should this be fixed on onnxscript setup side?

justinchuby replied:
I am not sure how this can be solved, since the package name is specified in the package meta.

justinchuby replied:
Perhaps install onnxscript without dependencies?

BowenBao replied:
yep, seems only way. I was wondering since people using onnx-weekly might run into this too.

BowenBao commented:
> Is the pass always turned on? Do we expect the current ort dev build to execute without errors when users want to test the new exporter?

Yes. Yes for PT 2.1, and no for now.

BowenBao committed:
Update on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

BowenBao committed:
update test env onnx installation script on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

BowenBao committed:
typo and nit on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+class Modularize(_pass.Transform):
+    """Transforms a flattened fx.GraphModule into a modular structure.
+
+    This method generates a new fx.GraphModule by creating fx submodules for all nodes
```
Kind of get lost in the whole paragraph. I thought fx submodules was sub fx.GraphModules, but now it seems like it's something different.

Looks like fx.GraphModule flatten the nested nn.Module in fx graph with logging them with node of `call_module`, and this pass is bringing back the nested structure to facilitate module as function exporting.

Is `call_module` only invokes by this pass now? I guess my question is that is `call_module` equal to `export module as function`?

BowenBao replied:
`call_module` is non-existent in previous fx graphs, (not considering fx cond). This pass creates `call_module` nodes, and the the corresponding sub `fx.GraphModule` as the node's target.

So in terms of relation with onnx, `call_module` node is converted to the function node in ONNX, and sub `fx.GraphModule` is converted to the function proto. This behavior is documented under `def call_module` in `fx_onnx_interpreter.py`.

wschin commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -634,9 +675,82 @@ def call_method(self, node: torch.fx.Node):
+        # module attribute name: `torch_nn_modules_conv_Conv2d_conv1`.
+        # Other naming styles such as qualified module class name made unique can also
+        # be considered.
+        unique_module_name = f"{sub_module._get_name()}_{node.target}"
```
Please add some tests to guard naming style. We want to have warning if any change affects the final node names.

BowenBao replied:
done

justinchuby commented on torch/onnx/_internal/exporter.py:
```diff
@@ -875,6 +875,14 @@ def pre_export_passes(
             diagnostic_context, module, original_model
         ).run()
 
+    # This operation should be invoked as the last pre export pass.
```
Should this note be in the class docstring as well?

BowenBao replied:
good idea

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+    """Meta information about a module.
+
+    This class is used to represent the module information in a more structured way.
+    It parses raw module information from node.meta.
```
```suggestion
    It parses raw module information from node.meta.
    See the uses of from_fx_tracer_produced_raw_meta and
    from_dynamo_produced_raw_meta for how to create an
    instance. 
```

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        return _ModuleMeta("", None, ("", None))
+
+    @classmethod
+    def from_fx_tracer_produced_raw_meta(
```
nit: Add tests for classmethod's.

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+
+
+class _ModuleStackMeta:
+    """Meta information about the module stack.
```
What is `module stack`? I guess it means the `model hierarchy` like
```
GPT(nn.Module)
  SelfAttentionLayer(nn.Module)
     Gemm
```
? But it could also mean the `call stack` invoked to reach a module's forward method. Could you make it clear? Thanks.

BowenBao replied:
name inherited from `nn_module_stack` meta. Updated docstring to make it more obvious.
And you are right, it is call stack.

Prefers current name, since it has matching raw meta.

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        if nn_module_stack_meta is None:
+            return
+        raw_meta = copy.copy(nn_module_stack_meta)
+        is_fx_tracer_produced_raw_meta = isinstance(
```
Move this logic to
```
class _ModuleMeta
    ...
    @classmethod
    def from_fx_tracer_produced_raw_meta(...):
        ...
     @classmethod
     def from_dynamo_produced_raw_meta(...):
        ...
     @classmethod
     def create(...):
         return from_fx_tracer_produced_raw_meta(...)
         or
         return from_dynamo_produced_raw_meta(...)
```

BowenBao replied:
great idea

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        return len(self._module_stack) == 0
+
+    def top(self) -> _ModuleMeta:
+        """Returns the top module meta in the stack. I.e., the meta for leaf module."""
```
The concept of `stack` is unclear to me. Could you elaborate in the doc? Thanks.

BowenBao replied:
added example

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        """Determines if self is a superset of the provided module stack.
+
+        I.e., If self includes all elements from the provided module stack, plus additional
+        elements op top. If self is empty or root, this method always return False.
```
```suggestion
        elements on top. If self is empty or root, this method always return False.
```
It's better to give one example here. There is no official definition of `module stack`.

BowenBao replied:
done

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+
+    @property
+    def raw_meta(self) -> Optional[Dict[str, Tuple[str, type]]]:
+        """Returns the raw module stack meta data, i.e. node.meta['nn_module_stack']."""
```
Ah, this is probably the most important word for defining `module stack`. Please add this in `_ModuleStackMeta`'s doc string and with an example of `print(node.meta["nn_module_stack"])`.

BowenBao replied:
done

BowenBao commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -171,6 +172,10 @@
     ),
+    xfail(
+        "all",
+        reason="[PostInline][ORT][ShapeInferenceError] axis must be in [-rank, rank-1]. input rank was 0"
```
Haven't looked closely, likely due to inliner and unrelated to module as function.
Suspect related to inlining exposing the inner if op that conditions on input rank.

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        return self.top().qualified_module_class_name
+
+
+@functools.lru_cache()
```
Let's not to use cache unless
1. it's a performance bottleneck.
2. we are 100% sure no modification on torch.fx.Node in torch.fx.GraphModule.

BowenBao replied:
ok

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        ...
+
+
+class _ModuleNode(_IRNode):
```
Wait? Another IR (Sorry. I am strongly against new IR)? Is FX IR not sufficient for modularization?

BowenBao replied:
Added new [NOTE: Modularize Pass Implementation] section. 

The IRNode is not as intimidating as the name suggests. `_LeafNode` and `_ModuleNode` are just natural abstractions to help with this pass and make code look more responable, as compared to creating a bunch of nested list/tuple of fx.Nodes otherwise.

wschin reviewed (dismissed):
Overall LGTM but I am concerning if we really need another IR to represent FX model. I assume FX IR can be reused for this work. Could you please try to reuse FX IR or ONNX IR?

Feel free to ping me if I don't respond on comments. I don't check github daily. Thanks.

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        >>> gm, _ = torch._dynamo.export(TestModule(), torch.tensor([0, 1, 2]), aten_graph=True)
+        >>> gm.print_readable()
+
+        >>> gm = passes.Modularize(infra.DiagnosticContext("test_context", "1.0"), gm).run()
```
This is great. Very short and reusable example. Thank you.

BowenBao replied:
:)

wschin commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        self, reference_root_module: torch.fx.GraphModule, stack_meta: _ModuleStackMeta
+    ):
+        self._stack_meta = stack_meta
+        self._nodes: List[_IRNode] = []
```
This looks like a `GraphModule` with extra metadata.

BowenBao replied:
mainly for a more natural representation of hierarchically grouping fxNodes, and then converting them into submodule.

BowenBao commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -485,12 +512,27 @@ def placeholder(
+            output = onnxscript_graph.add_input(
+                input_name=node.name,
+                shape=[],
+                dtype=sym_type_to_tensor_type[type(fake_tensor)],
```
FLOAT[] for example. A scalar tensor.

BowenBao commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -485,12 +512,27 @@ def placeholder(
                 dtype=fake_tensor.dtype,
             )
+        elif isinstance(fake_tensor, (torch.SymBool, torch.SymInt, torch.SymFloat)):
+            sym_type_to_tensor_type = {
```
good idea

BowenBao commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -188,6 +189,16 @@ def filter_incompatible_and_dtype_convert_kwargs(kwargs):
+    torch.SymInt: torch.int64,
+    torch.SymFloat: torch.float32,
+    torch.SymBool: torch.bool,
+}
```
sure

BowenBao commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        self._module_stack.append(module_meta)
+
+    @_beartype.beartype
+    def is_same(
```
removed, thx

thiagocrepaldi reviewed (commented):
I understood the ultimate goal, especially with the examples and tests.

We probably would benefit on a high level view on how l new concepts such as `_moduleMeta`, `ModuleStackMeta`, `_IRNode`, `_ModuleNode`,`_LeafNode` come together to extend the existing FX

The dynamo/fx symbolic differences are not clear. Maybe it could be refactored to be more modularized too

BowenBao committed:
Add more comments about impl on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+class _IRNode(abc.ABC):
+    """Base class for IR nodes.
+
+    IR nodes are used for modularization. They add a layer of abstraction on top of
```
https://github.com/pytorch/pytorch/pull/105618/files#r1270159849

BowenBao commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        return f"ModuleMeta(name={self._module_name}, class={self._module_class})"
+
+    @classmethod
+    def root(cls) -> _ModuleMeta:
```
ok

BowenBao commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -224,6 +233,8 @@ def _fill_tensor_shape_type(
             onnxscript_value.dtype = torch.int64
         elif isinstance(expected_value, torch.SymFloat):
             onnxscript_value.dtype = torch.float32
+        elif isinstance(expected_value, torch.SymBool):
```
Not sure. Maybe more cases will show as we run bench with dynamic.

BowenBao commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -590,6 +599,7 @@ def test_output_match(self, device: str, dtype: torch.dtype, op):
                 inputs=repr(inputs),
                 kwargs=repr(cpu_sample.kwargs),
             ):
+                print("inputs: ", inputs, " kwargs: ", cpu_sample.kwargs)
```
yea not sure where they are from. thx

BowenBao commented on test/onnx/onnx_test_common.py:
```diff
@@ -312,6 +312,24 @@ def run_ort(
+    else:
+        if isinstance(ort_model, bytes):
+            buffer = io.BytesIO(ort_model)
+        else:
```
good catch, it seems redundant.

BowenBao replied:
no it is not redundant. I can add assert if you'd like but i feel it is dup checking since onnx_model is type annotated and this function is guarded by beartype.

BowenBao commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        >>> gm.print_readable()
+
+        >>> gm = passes.Modularize(infra.DiagnosticContext("test_context", "1.0"), gm).run()
+        >>> gm.print_readable()
```
I actually removed the outputs after writing it up. It feels too verbose and it includes volatile info that might cause this doctest to break often. Since this is doctest guarded by CI, the snippet is always guaranteed to work so people can easily copy paste and run. I think this is good enough balance?

BowenBao commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+    The third interpretation is the most specific scenario. It necessitates the minimum
+    analysis and creates the most stable representation. The drawback is that it
+    generates more redundancy than the other two methods. If needed, a subsequent post-processing
+    pass can be applied to consolidate completely identical functions and reduce duplication.
```
Yea this is discussing high level choice of what each submodule represents. I have added a few lines to explain on high level how this pass works, and link to docstring of more details.

BowenBao commented on torch/onnx/_internal/fx/passes/modularization.py:
```diff
@@ -0,0 +1,648 @@
+        return _ModuleMeta("", None, ("", None))
+
+    @classmethod
+    def from_fx_tracer_produced_raw_meta(
```
This is the helper function for the one `from_raw_meta` function (at the point of your comment it is from the init function of another class). `from_raw_meta` does not need to know which tracer produces the raw data, it parses the raw data and decides what to do. The naming is simply like a comment hinting it is the pattern from fx_tracer.

BowenBao committed:
comment on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

BowenBao committed:
unittest to cover function names on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

BowenBao commented on test/onnx/test_fx_passes.py:
```diff
@@ -101,6 +101,76 @@ def func(x, y, z):
+                out = self.relu(out)
+                return out
+
+        torch.onnx.dynamo_export(TestModule(), torch.randn(3), torch.randn(3))
```
done

BowenBao commented on test/onnx/test_fx_passes.py:
```diff
@@ -101,6 +101,76 @@ def func(x, y, z):
+                return out
+
+        torch.onnx.dynamo_export(TestModule(), torch.randn(3), torch.randn(3))
+
```
done

BowenBao committed:
typo and adjustment on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

BowenBao committed:
stablize test on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

BowenBao committed:
discuss why not write as interpreter on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

BowenBao committed:
lint on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

justinchuby commented on .ci/docker/common/install_onnx.sh:
```diff
@@ -24,7 +28,7 @@ pip_install \
 
 # TODO: change this when onnx-script is on testPypi
-pip_install "onnxscript@git+https://github.com/microsoft/onnxscript@2bb3e9f2d094912f81cb63cecb412efb14c65738"
+pip_install "onnxscript@git+https://github.com/microsoft/onnxscript@2bb3e9f2d094912f81cb63cecb412efb14c65738" --no-deps
```
fyi https://pypi.org/project/onnxscript-preview now available

BowenBao replied:
What is versioning for preview btw? is it like by dates like the nightly s?

justinchuby replied:
Yes. It's a two-times-a-week release version by dates

BowenBao committed:
rebase and use onnxscript-preview on "[ONNX] Export module as function"


Introduce `Modularize` pass that analyzes the flat `fx.GraphModule` and creates nested
layers of sub `fx.GraphModule`s along with the `call_module` fx nodes that invokes them.
The analysis is done on the meta data "nn_module_stack", which captures the `nn.Module`
each flat `fx.Node` belongs to.

`FxOnnxInterpreter` is updated to support `call_module`. The related sub module linked
by `node.target` is exported as an ONNX model local function. The `call_module` node itself
is exported as an ONNX node, associated with the ONNX model local function by op_type.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge
